platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -137,6 +154,280 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	armv8_cipher_constants_t cc = {
+		.mode = 0
+	};
+	uint8_t *iv_ptr;
+	uint64_t nonce_length, nonce_byte_length;
+	uint64_t plaintext_length, plaintext_byte_length;
+	uint64_t aad_length, aad_byte_length;
+	int ret = 0;
+
+	/* Clear the allocated buffers */
+	memset(global->input, 0, _ARM_CRYPTO_MAX_DATA_LENGTH);
+	memset(global->nonce, 0, _ARM_CRYPTO_MAX_IV_LENGTH);
+	memset(global->aad, 0, _ARM_CRYPTO_MAX_AAD_LENGTH);
+	memset(global->digest, 0, _ARM_CRYPTO_MAX_DIGEST_LENGTH);
+	memset(global->output, 0, (_ARM_CRYPTO_MAX_DATA_LENGTH +
+	       _ARM_CRYPTO_MAX_DIGEST_LENGTH));
+
+	nonce_length = (session->p.cipher_iv.length) * 8;
+	/* IV length must be a multiple of 16 */
+	nonce_byte_length = ((nonce_length + 127) & ~127UL) >> 3;
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	/* Copy nonce and pad it to nonce_byte_length */
+	for (uint64_t i = 0; i < session->p.cipher_iv.length; i++)
+		global->nonce[i] = iv_ptr[i];
+	for (uint64_t i = session->p.cipher_iv.length; i < nonce_byte_length; i++)
+		global->nonce[i] = 0xde;
+
+	plaintext_length = (param->cipher_range.length) * 8;
+	/* Plaintext length must be a multiple of 16 */
+	plaintext_byte_length = ((plaintext_length + 127) & ~127UL) >> 3;
+
+	/* Copy plaintext and pad it to plaintext_byte_length */
+	uint32_t offset = param->cipher_range.offset;
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, offset, &seg_len, NULL);
+
+	for (uint64_t i = 0; i < param->cipher_range.length; i++)
+		global->input[i] = data[i];
+	for (uint64_t i = param->cipher_range.length; i < plaintext_byte_length; i++)
+		global->input[i] = 0xde;
+
+	aad_length = (session->p.auth_aad_len) * 8;
+	/* Pad AAD length to multiple of 16 */
+	aad_byte_length = ((aad_length + 127) & ~127UL) >> 3;
+
+	/* Copy AAD and pad it to aad_byte_length */
+	for (uint64_t i = 0; i < session->p.auth_aad_len; i++)
+		global->aad[i] = param->aad_ptr[i];
+	for (uint64_t i = session->p.auth_aad_len; i < aad_byte_length; i++)
+		global->aad[i] = 0xde;
+
+	cs.constants = &cc;
+	switch (session->p.cipher_key.length) {
+	case 16:
```

Comment:
This sort of testing does not need to be done for every packet. Session creation can do all setup that does not depend on the packet or packet specific parameters.

Reply:
I am not sure I understand what the issue is here. According to the cipher key length (128/192/256 bits), the cipher mode needs to be set in struct armv8_cipher_constants_t. I agree this is not a packet specific parameter but if this to be moved to session creation, then from my understanding we have to make the following change:-
Add a member to struct odp_crypto_generic_session_t (or some other struct) to add a member and set that structure member so that the mode setting is passed back to encrypt/decrypt function. 

Is this solution acceptable? Please suggest any other possible options for this which may work better

Reply:
> I am not sure I understand what the issue is here.

For every packet encrypted you are redoing calculations that stay valid over the life of the session. It is not only the switch statement figuring out the cipher, but also the subsequent call to armv8_aes_gcm_set_constants(), which performs AES key expansion. Things that stay constant for a session should be set up at session creation time and stored in the session so that the per-packet encryption function does not have to redo them over and over again.

What I am saying is that not only the "mode" but also the cipher constants that contain the expanded key should be stored in session struct.

Reply:
Okay, understood. I'll work on moving these lines of code to the session creation and make some changes to struct odp_crypto_generic_session_t